File: src/cmake.rs
// expose cmake utilities

// subprocess
// typing
// shutil
// os
// shlex
// core
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::
{
    core,
    cmd,
    found,
    registry,
    printer
};


fn find_cmake_in_registry(printer: &mut printer::Printer) -> found::Found
{
    let registry_source = "registry".to_string();

    match registry::hklm(r"SOFTWARE\Kitware\CMake", "InstallDir")
    {
        Err(_) => found::Found::new(None, registry_source),
        Ok(install_dir) =>
        {
            let bpath: PathBuf = [install_dir.as_str(), "bin", "cmake.exe"].iter().collect();
            let spath = bpath.as_path();
            let path = spath.to_str().unwrap();
            if spath.exists()
            {
                found::Found::new(Some(path.to_string()), registry_source)
            }
            else
            {
                printer.error(format!("Found path to cmake in registry ({}) but it didn't exist", path).as_str());
                found::Found::new(None, registry_source)
            }
        }
    }
}


fn find_cmake_in_path(printer: &mut printer::Printer) -> found::Found
{
    let path_source = "path".to_string();
    match which::which("cmake")
    {
        Err(_) => found::Found::new(None, path_source),
        Ok(bpath) =>
        {
            let spath = bpath.as_path();
            let path = spath.to_str().unwrap();
            if spath.exists()
            {
                found::Found::new(Some(path.to_string()), path_source)
            }
            else
            {
                printer.error(format!("Found path to cmake in path ({}) but it didn't exist", path).as_str());
                found::Found::new(None, path_source)
            }
        }
    }
}


pub fn list_all(printer: &mut printer::Printer) -> Vec::<found::Found>
{
    vec![find_cmake_in_registry(printer), find_cmake_in_path(printer)]
}


fn find_cmake_executable(printer: &mut printer::Printer) -> Option<String>
{
    let list = list_all(printer);
    found::first_value_or_none(&list)
}


// a cmake argument
struct Argument
{
    name: String,
    value: String,
    typename: Option<String>
}

impl Argument
{
    pub fn new(name: String, value: String, typename: Option<String>) -> Argument
    {
        Argument
        {
            name,
            value,
            typename
        }
    }

    // format for commandline
    pub fn format_cmake_argument(&self) -> String
    {
        match &self.typename
        {
            Some(t) => format!("-D{}:{}={}", self.name, t, self.value),
            None => format!("-D{}={}", self.name, self.value)
        }
    }
}


// cmake generator
pub struct Generator
{
    generator: String,
    arch: Option<String>
}

impl Generator
{
    pub fn new_with_arch(generator: &str, arch: &str) -> Generator
    {
        Generator
        {
            generator: generator.to_string(),
            arch: Some(arch.to_string())
        }
    }
    pub fn new(generator: &str) -> Generator
    {
        Generator
        {
            generator: generator.to_string(),
            arch: None
        }
    }
}


// utility to call cmake commands on a project
pub struct CMake
{
    generator: Generator,
    build_folder: PathBuf,
    source_folder: PathBuf,
    arguments: Vec::<Argument>
}


impl CMake
{
    pub fn new(build_folder: &Path, source_folder: &Path, generator: Generator) -> CMake
    {
        CMake
        {
            generator,
            build_folder: build_folder.to_path_buf(),
            source_folder: source_folder.to_path_buf(),
            arguments: Vec::<Argument>::new()
        }
    }

    // add argument with a explicit type set
    pub fn add_argument_with_type(&mut self, name: String, value: String, typename: String)
    {
        self.arguments.push(Argument::new(name, value, Some(typename)))
    }

    // add argument
    pub fn add_argument(&mut self, name: String, value: String)
    {
        self.arguments.push(Argument::new(name, value, None))
    }

    // set the install folder
    pub fn set_install_folder(&mut self, folder: &Path)
    {
        self.add_argument_with_type("CMAKE_INSTALL_PREFIX".to_string(), folder.to_string_lossy().to_string(), "PATH".to_string())
    }

    // set cmake to make static (not shared) library
    pub fn make_static_library(&mut self)
    {
        self.add_argument("BUILD_SHARED_LIBS".to_string(), "0".to_string())
    }

    // run cmake configure step
    pub fn config(&self, printer: &mut printer::Printer) { self.config_with_print(printer, false); }
    pub fn config_with_print(&self, printer: &mut printer::Printer, only_print: bool)
    {
        let found_cmake = find_cmake_executable(printer);
        let cmake = match found_cmake
        {
            Some(f) => {f},
            None =>
            {
                printer.error("CMake executable not found");
                return;
            }
        };
        
        let mut command = Command::new(cmake);
        for arg in &self.arguments
        {
            let argument = arg.format_cmake_argument();
            printer.info(format!("Setting CMake argument for config: {}", argument).as_str());
            command.arg(argument);
        }

        command.arg(self.source_folder.to_string_lossy().to_string());
        command.arg("-G");
        command.arg(self.generator.generator.as_str());
        match &self.generator.arch
        {
            Some(arch) =>
            {
                command.arg("-A");
                command.arg(arch);
            }
            None => {}
        }
        
        core::verify_dir_exist(&self.build_folder);
        command.current_dir(self.build_folder.to_string_lossy().to_string());
        
        if core::is_windows()
        {
            if only_print
            {
                printer.info(format!("Configuring cmake: {:?}", command).as_str());
            }
            else
            {
                // core::flush();
                cmd::check_call(printer, &mut command);
            }
        }
        else
        {
            printer.info(format!("Configuring cmake: {:?}", command).as_str());
        }
    }

    // run cmake build step
    pub fn build_cmd(&self, printer: &mut printer::Printer, install: bool)
    {
        let found_cmake = find_cmake_executable(printer);
        let cmake = match found_cmake
        {
            Some(f) => {f},
            None =>
            {
                printer.error("CMake executable not found");
                return;
            }
        };

        let mut command = Command::new(cmake);
        command.arg("--build");
        command.arg(".");

        if install
        {
            command.arg("--target");
            command.arg("install");
        }
        command.arg("--config");
        command.arg("Release");

        core::verify_dir_exist(&self.build_folder);
        command.current_dir(self.build_folder.to_string_lossy().to_string());

        if core::is_windows()
        {
            // core::flush()
            cmd::check_call(printer, &mut command);
        }
        else
        {
            printer.info(format!("Calling build on cmake: {:?}", command).as_str());
        }
    }

    // build cmake project
    pub fn build(&self, printer: &mut printer::Printer)
    {
        self.build_cmd(printer, false);
    }

    // install cmake project
    pub fn install(&self, printer: &mut printer::Printer)
    {
        self.build_cmd(printer, true);
    }
}

File: src/CMake.cs
using Newtonsoft.Json;
using System.Runtime.Serialization;

namespace Workbench.CMake;


static class CmakeTools
{
    public static Found find_cmake_in_registry(Printer printer)
    {
        var registry_source = "registry";

        var install_dir = Registry.hklm(@"SOFTWARE\Kitware\CMake", "InstallDir");
        if (install_dir == null) { return new Found(null, registry_source); }

        var path = Path.Join(install_dir, "bin", "cmake.exe");
        if (File.Exists(path) == false)
        {
            printer.error($"Found path to cmake in registry ({path}) but it didn't exist");
            return new Found(null, registry_source);
        }

        return new Found(path, registry_source);
    }


    public static Found find_cmake_in_path(Printer printer)
    {
        var path_source = "path";
        var path = Which.Find("cmake");
        if (path == null)
        {
            return new Found(null, path_source);
        }

        if (File.Exists(path) == false)
        {
            printer.error($"Found path to cmake in path ({path}) but it didn't exist");
            return new Found(null, path_source);
        }
        return new Found(path, path_source);
    }


    public static IEnumerable<Found> list_all(Printer printer)
    {
        yield return find_cmake_in_registry(printer);
        yield return find_cmake_in_path(printer);
    }


    public static string? find_cmake_executable(Printer printer)
    {
        return Found.first_value_or_none(list_all(printer));
    }
}

// a cmake argument
public class Argument
{
    public Argument(string name, string value)
    {
        this.name = name;
        this.value = value;
    }

    public Argument(string name, string value, string typename) : this(name, value)
    {
        this.typename = typename;
    }

    // format for commandline
    public string format_cmake_argument()
    {
        if (typename == null)
        {
            return $"-D{this.name}={this.value}";
        }
        else
        {
            return $"-D{this.name}:{typename}={this.value}";
        }
    }

    public string name { get; }
    public string value { get; }
    public string? typename { get; }
}


// cmake generator
public class Generator
{
    public Generator(string generator, string? arch = null)
    {
        this.generator = generator;
        this.arch = arch;
    }

    public string generator { get; }
    public string? arch { get; }
}

// utility to call cmake commands on a project
public class CMake
{
    public CMake(string build_folder, string source_folder, Generator generator)
    {
        this.generator = generator;
        this.buildFolder = build_folder;
        this.sourceFolder = source_folder;
    }

    private readonly Generator generator;
    private readonly string buildFolder;
    private readonly string sourceFolder;
    private readonly List<Argument> arguments = new List<Argument>();


    // add argument with a explicit type set
    private void AddArgumentWithType(string name, string value, string typename)
    {
        this.arguments.Add(new Argument(name, value, typename));
    }

    // add argument
    public void AddArgument(string name, string value)
    {
        this.arguments.Add(new Argument(name, value));
    }

    // set the install folder
    public void SetInstallFolder(string folder)
    {
        this.AddArgumentWithType("CMAKE_INSTALL_PREFIX", folder, "PATH");
    }

    // set cmake to make static (not shared) library
    public void MakeStaticLibrary()
    {
        this.AddArgument("BUILD_SHARED_LIBS", "0");
    }

    // run cmake configure step
    public void Configure(Printer printer, bool nop = false)
    {
        var cmake = CmakeTools.find_cmake_executable(printer);
        if (cmake == null)
        {
            printer.error("CMake executable not found");
            return;
        }

        var command = new Command(cmake);
        foreach (var arg in this.arguments)
        {
            var argument = arg.format_cmake_argument();
            printer.Info($"Setting CMake argument for config: {argument}");
            command.AddArgument(argument);
        }

        command.AddArgument(this.sourceFolder);
        command.AddArgument("-G");
        command.AddArgument(this.generator.generator);
        if (generator.arch != null)
        {
            command.AddArgument("-A");
            command.AddArgument(generator.arch);
        }

        Core.VerifyDirectoryExists(printer, this.buildFolder);
        command.WorkingDirectory = this.buildFolder;

        if (Core.IsWindows())
        {
            if (nop)
            {
                printer.Info($"Configuring cmake: {command}");
            }
            else
            {
                command.RunAndGetOutput().PrintStatusAndUpdate(printer);
            }
        }
        else
        {
            printer.Info($"Configuring cmake: {command}");
        }
    }

    // run cmake build step
    private void RunBuildCommand(Printer printer, bool install)
    {
        var cmake = CmakeTools.find_cmake_executable(printer);
        if (cmake == null)
        {
            printer.error("CMake executable not found");
            return;
        }

        var command = new Command(cmake);
        command.AddArgument("--build");
        command.AddArgument(".");

        if (install)
        {
            command.AddArgument("--target");
            command.AddArgument("install");
        }
        command.AddArgument("--config");
        command.AddArgument("Release");

        Core.VerifyDirectoryExists(printer, this.buildFolder);
        command.WorkingDirectory = this.buildFolder;

        if (Core.IsWindows())
        {
            command.RunAndGetOutput().PrintStatusAndUpdate(printer);
        }
        else
        {
            printer.Info($"Found path to cmake in path ({command}) but it didn't exist");
        }
    }

    // build cmake project
    public void Build(Printer printer)
    {
        this.RunBuildCommand(printer, false);
    }

    // install cmake project
    public void Install(Printer printer)
    {
        this.RunBuildCommand(printer, true);
    }
}

[JsonObject(MemberSerialization.OptIn)]
public class Trace
{
    public Trace(string file, int line, string cmd, string[] args)
    {
        this.File = file;
        this.Line = line;
        this.Cmd = cmd;
        this.Args = args;
    }

    [JsonProperty("file")]
    public string File { get; }

    [JsonProperty("line")]
    public int Line { get; }

    [JsonProperty("cmd")]
    public string Cmd { get; }

    [JsonProperty("args")]
    public string[] Args { get; }

    public static IEnumerable<Trace> TraceDirectory(string dir)
    {
        List<Trace> lines = new();
        List<string> error = new();

        void on_line(string src)
        {
            try
            {
                var parsed = JsonConvert.DeserializeObject<Trace>(src);
                if (parsed != null && parsed.File != null)
                {
                    // file != null ignores the version json object
                    lines.Add(parsed);
                    return;
                }
            }
            catch (Newtonsoft.Json.JsonReaderException)
            {
                // pass
            }

            error.Add(src);
        }

        var ret = new Command("cmake", "--trace-format=json-v1")
            .InDirectory(dir)
            .RunWithCallback(on_line)
            .ExitCode
            ;

        if (ret != 0)
        {
            var mess = string.Join('\n', error);
            throw new TraceError($"{error.Count} -> {mess}");
        }

        return lines;
    }
}

[Serializable]
internal class TraceError : Exception
{
    public TraceError()
    {
    }

    public TraceError(string message) : base(message)
    {
    }

    public TraceError(string? message, Exception? innerException) : base(message, innerException)
    {
    }

    protected TraceError(SerializationInfo info, StreamingContext context) : base(info, context)
    {
    }
}

